### PR TITLE
Support BigQuery OAuth using a refresh token and client secrets

### LINKED ISSUE #2344: Support BigQuery oauth using a bearer token
### Describe the feature
dbt Core should support BigQuery targets configured directly with a Bearer token for OAuth authentication. This will enable OAuth authentication without requiring manual `gcloud` authorization steps to log into GCP.

This feature should look similar to the Snowflake OAuth flow implemented in #2050. Let's create a new auth method called `bearer` which accepts the following fields:
 - `client_id`
 - `client_secret`
 - `refresh_token`
 -  `token_uri`

Given these credentials, dbt should be able to mint new access tokens that are used to authorize BigQuery queries.

### Who will this benefit?
Services that invoke dbt on behalf of a user (ie. where Service Account auth isn't desirable). We plan on leveraging this functionality in dbt Cloud to support OAuth dev creds for BigQuery credentials, but anyone should be able to use this API too.

### LINKED ISSUE #2802: [BigQuery] Raw token authentication method
### Describe the feature
The existing auth methods for BigQuery provide a great experience for interactive users, allowing them to transparently provision access tokens in a variety of ways.  However, as an automation implementer on a team trying to programmatically wield dbt on behalf of customers, I'd like to be able to bypass these niceties and just inject an externally provisioned token for access.

In our typical non-dbt scenarios, we create service accounts for customers, have them grant specific limited permissions to those accounts for our service's operations, then use our master account to issue scoped tokens for those accounts/operations when they need to execute.  When running dbt operations, I'd like to avoid writing our master account credentials file to disk (as required by the `service-account`/`service-account-json` methods) to protect against reflected file attacks and/or potential vulnerabilities in dbt itself, which could potentially exfiltrate these creds.  While I don't see these as particularly likely scenarios, I'd love to have the ability to directly control the blast radius via smaller scoped credentials.

I propose a new BigQuery auth method named 'service-token' and an additional field named `service_token` to provide its value, e.g.:
```
my-bigquery-db:
  target: dev
  outputs:
    dev:
      type: bigquery

      # something like this?
      method: service-token
      service_token: 'ya29.c.KqwC3gfRlXO...'

      project: [GCP project id]
      dataset: [the name of your dbt dataset]
      threads: [1 or more]
      timeout_seconds: 300
      priority: interactive
      retries: 1
```

### Describe alternatives you've considered
With the addition of the `impersonate_service_account` field added in 0.18.0, I can connect as delegated services, but still need to provide the full master service keyfile rather than a more limited credential.

### Additional context
This is specific to BigQuery database connections and automation use-cases.  I think it's fair to say that standard interactive users would almost never want to use this authentication mode directly.

### Who will this benefit?
* Security-minded SaaS automators who are targeting multi-warehouse and connecting to BQ on behalf of customers
* IT/Dataeng staff using automation to provision limited credentials as part of employee access control

### Are you interested in contributing this feature?
Yes, happy to contribute to making this a reality.  I have a simple (if slightly hacky) proof-of-concept that I can polish into a real PR if this seems interesting.

### PR DESCRIPTION
**Pull request by drewbanin**
resolves #2344
resolves #2802

### Description

This PR adds support for the `bearer` connection method on BQ. When this method is specified, a `client_id`, `client_secret`, `refresh_token` and `token_uri` must also be provided. Given these secrets, dbt will mint access tokens to authorize BigQuery queries.

### TODO
 - [x] Should we rename `bearer` to `oauth-secrets` as the auth method?
 - [x] Write tests
 - [x] Changelog entry

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: plugins/bigquery/dbt/adapters/bigquery/connections.py [hunk 1]
```diff
@@ -165,6 +179,16 @@ def get_bigquery_credentials(cls, profile_credentials):
             details = profile_credentials.keyfile_json
             return creds.from_service_account_info(details, scopes=cls.SCOPE)
 
+        elif method == BigQueryConnectionMethod.BEARER:
```
Comment (drewbanin) [inline plugins/bigquery/dbt/adapters/bigquery/connections.py:182]: @jtcohen6 do you have any thoughts / opinions on what we call this auth method?

Comment (jtcohen6) [reply plugins/bigquery/dbt/adapters/bigquery/connections.py:182]: I don't mind `bearer` as long as it reflects what BQ calls this ([right?](https://cloud.google.com/bigquery/docs/authorization)). I prefer it to `oauth-secrets`

Comment (drewbanin) [reply plugins/bigquery/dbt/adapters/bigquery/connections.py:182]: ^ it's actually different than the method described in that link, which is kind of why i _do not_ want to call this bearer! Bearer auth actually feels more to me like the change described in https://github.com/fishtown-analytics/dbt/issues/2802 :D

Comment (jtcohen6) [reply plugins/bigquery/dbt/adapters/bigquery/connections.py:182]: QED. Let's call this something else. `oauth-refresh-token`?

**Comment by heisencoder** [comment]
I'm pretty excited about this feature!  Let me know if you need any help with an integration test case (whether that's feasible).  Currently, the BIGQUERY_SERVICE_ACCOUNT_JSON is passed in as an environment variable, so maybe a similar variable could be something like BIGQUERY_REFRESH_TOKEN_JSON, which would contain the parameters to the GoogleCredentials.Credentials method.

File: plugins/bigquery/dbt/adapters/bigquery/connections.py [hunk 2]
```diff
@@ -51,19 +54,30 @@ class BigQueryConnectionMethod(StrEnum):
     OAUTH = 'oauth'
     SERVICE_ACCOUNT = 'service-account'
     SERVICE_ACCOUNT_JSON = 'service-account-json'
+    OAUTH_SECRETS = 'oauth-secrets'
```
Comment (jtcohen6) [inline plugins/bigquery/dbt/adapters/bigquery/connections.py:57]: Were you going to rename this to `refresh-token`?

Comment (drewbanin) [reply plugins/bigquery/dbt/adapters/bigquery/connections.py:57]: i decided not to! I actually updated this PR to also accept an access token directly via the `token` configuration. So, it's kind of just a grab bag of oauth auth methods. Google lets you supply:
 - a client id / client secret / refresh token / token uri OR
 - a standalone access token

The access token is going to expire after ~60 mins (not well defined), but it was a one-line change and feels like a reasonable enough use case to support. You buy it?

Comment (jtcohen6) [reply plugins/bigquery/dbt/adapters/bigquery/connections.py:57]: gotcha — does that mean that this resolves #2802 as well?

**Review by jtcohen6 — APPROVED** [review]
neat!

**Review by gshank — APPROVED** [review]

**Pull request merged**